drivers/spi/spi_bitbang.c

```
@@ -0,0 +1,293 @@
+
+			if (!loop && do_read && cpha) {
+				b = gpio_pin_get_dt(&info->miso_gpio);
+			}
+
+			if (loop) {
+				b = d;
+			}
+
+			r = (r << 1) | (b ? 0x1 : 0x0);
+
+			--shift;
+		}
+
+		if (spi_context_rx_buf_on(ctx)) {
+			switch (data->dfs) {
+			case 2:
+				*(uint16_t *)(ctx->rx_buf) = r;
+				break;
+			case 1:
+				*(uint8_t *)(ctx->rx_buf) = r;
+				break;
+			}
+		}
+
+		LOG_DBG(" w: %04x, r: %04x , do_read: %d", w, r, do_read);
+
+		spi_context_update_tx(ctx, data->dfs, 1);
+		spi_context_update_rx(ctx, data->dfs, 1);
+	}
+
+	spi_context_cs_control(ctx, false);
+
+	spi_context_complete(ctx, 0);
+
+	return 0;
+}
+
+#ifdef CONFIG_SPI_ASYNC
+static int spi_bitbang_transceive_async(const struct device *dev,
+				    const struct spi_config *spi_cfg,
+				    const struct spi_buf_set *tx_bufs,
+				    const struct spi_buf_set *rx_bufs,
+				    struct k_poll_signal *async)
+{
+	return -ENOTSUP;
+}
+#endif
+
+int spi_bitbang_release(const struct device *dev,
+			  const struct spi_config *config)
+{
+	struct spi_bitbang_data *data = dev->data;
+	struct spi_context *ctx = &data->ctx;
+
+	spi_context_unlock_unconditionally(ctx);
+	return 0;
+}
+
+static struct spi_driver_api spi_bitbang_api = {
+	.transceive = spi_bitbang_transceive,
+	.release = spi_bitbang_release,
+#ifdef CONFIG_SPI_ASYNC
+	.transceive_async = spi_bitbang_transceive_async,
+#endif /* CONFIG_SPI_ASYNC */
+};
+
+int spi_bitbang_init(const struct device *dev)
+{
+	const struct spi_bitbang_config *config = dev->config;
+
+	if (!device_is_ready(config->clk_gpio.port)) {
+		LOG_ERR("GPIO port for clk pin is not ready");
+		return -ENODEV;
+	}
+	if (gpio_pin_configure_dt(&config->clk_gpio, GPIO_OUTPUT_INACTIVE)) {
+		LOG_ERR("Couldn't configure clk pin");
+		return -EIO;
+	}
+
+	if (!device_is_ready(config->mosi_gpio.port)) {
+		LOG_ERR("GPIO port for mosi pin is not ready");
+		return -ENODEV;
+	}
+	if (gpio_pin_configure_dt(&config->mosi_gpio, GPIO_OUTPUT_INACTIVE)) {
```

Please propagate original errors to the callers instead of swallowing and returning -EIO, here and elsewhere.